### PR TITLE
feat(ecam): Add Ecam Warning - AP, A/THR DISC, L/G NOT DOWN

### LINKED ISSUE #1503: [REQUEST] Autopilot disconnect sound.
**Is your feature request related to a problem? Please describe.**

When disconnect autothrust, there is a no sound and no warning about disconnect autothrust

If we press autothrottle disconnect button on thurst lever's red button, a320neo should beep 'warning sound' once, and beep 3 times. 

**Describe the feature you'd like to see implemented**

**References**

https://youtu.be/Vq8Lj0pp39o?t=417
you can watch when disconnect autothurst

**Additional context**

Discord username (if different from GitHub):

### LINKED ISSUE #3727: [BUG] Master Alarm for TakeOff with Parking Brake on cannot be switched off
**Mod Version**
  "built": "2021-02-25T21:43:01+00:00",
  "ref": "master",
  "sha": "472cf527726e63d24e95608b6ac66d3430ffaacb",
  "actor": "Benjozork",
  "event_name": "manual"

**Describe the bug**
The Master Caution Alert cannot be switched off, when on a Runway, setting Thrust to TakeOff and having the Parking Brake still set to on.

**To Reproduce**
1. Load a new flight on any runway at any airport
2. Set Thrust to TakeOff or Max and leave the Parking Brake on
3. Try to disable the alarm - it won't switch off, even if you set the parking brake to off and try again.

**Expected behavior**
Master alarm can be cleared and alarm sound disappears

**Actual behavior**
Master alarm remains on all the time

**References**
- none

**Additional context**

Was this working before/when did the issue start occurring?
Haven't tried this for a very long time, but it used to work fine.

Is this a problem in the vanilla unmodded game?
No, a master alarm does not appear in the unmodded game at all.

### LINKED ISSUE #3888: [BUG] Master caution light & aural warning can’t be disabled after Sim Update 3, MSFS Version 1.14.5.0
**Mod Version**

I am using the Development Version
"built": "2021-03-10T00:43:09+00:00",
"ref": "master",
"sha": "f26a1291f98e395d85427fa4caf9414ef663e0ca",
"actor": "Benjozork",
"event_name": "manual"

**Describe the bug**

Master caution light & aural warning can’t be disabled. I got a master caution and couldn’t disable/cancel it after I pressed the master caution button.

**To Reproduce**

1.   Update sim to latest version and cause a master caution warning, then try to cancel it by pressing the master caustion button.

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #3727
Fixes #1503
Fixes #3888 

## Summary of Changes

when Autopilot & Autothrottle disconnected, ECAM warning & aural warning will ringing.

Also add some master warning / master caution

## References

image
image
image
image
image
image
image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

1. Spawn on runway or c&d as you wish
2. takeoff and "should" climb more than 1500ft
3. fly with autopilot & athr
4. disconnect autopilot & watch Upper ECAM & Master warning button
5. disconnect a/thr and watch UpperECAM & master caution, and play beep
6. if you check 1~5, next is l/g not down.
7. there is a 3 condition about l/g not down
7-1. Radio height lower than 750ft without gear downlocked(gear downlocked mean gear is fully down) / both engine's N1 lower than 75%
7-2. on 7-1 situation, if some engine is shutdown, other engine's N1 lower than 97%
7-3. In 7-1 situation, if flaps not retracted(so flap setting is 1 2 3 or full), if engine is not TOGA, alert will appear

note : you should clear of conflict L/G NOT DOWN first with gear down or climb more than RADIO HEIGHT 750ft and again test.
because current we can't use EMER CANC.

WARNING
1. L/G NOT DOWN CAN NOT CANCEL BY MASTER WARNING
    - please use under ECAM, CLR button
2. Correct behavior is can't cancel with CLR button before FWC Rewrite.

check all 7-1~3 situation please

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 1]
```diff
@@ -401,4 +416,78 @@ class A32NX_FWC {
         }
         return true;
     }
+
+    _autopilotDisconnect(_deltaTime) {
+        const apStatus = SimVar.GetSimVarValue("AUTOPILOT MASTER", "Bool");
+        const atherStatus = SimVar.GetSimVarValue("AUTOTHROTTLE ACTIVE", "Bool");
```
Comment (beheh) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:422]: What's `ather`? Do you mean `athr`, or `autothrottle`?

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 2]
```diff
@@ -401,4 +416,78 @@ class A32NX_FWC {
...
+            if (this.warningPressed === true || (this.apdeltaTime / 1000) >= 3) {
+                this.AutopilotWarningCanceled = true;
+                SimVar.SetSimVarValue("L:A32NX_AP_DISC", "Bool", false);
+                SimVar.SetSimVarValue("L:Generic_Master_Warning_Active", "Bool", false);
+                this.apdeltaTime = 0;
+            }
+        }
+    }
+
+    _checkLandingGear(_deltaTime) {
```
Comment (beheh) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:463]: This does not need to be run every single tick. especially as it always access all these SimVars. How about we use the update throttler class to only run these specific checks once or twice per second?

**Review by beheh — CHANGES_REQUESTED** [review]

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 2 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:463]: Ok!

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 1 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:422]: Yes i'll fix it

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 2 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:469]: I added it but idk that is correct. can you check it?

Comment (beheh) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:469]: That looks correct! Does it still work?

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 3]
```diff
@@ -76,7 +85,12 @@ class A32NX_FWC {
         this._updateButtons(_deltaTime);
         this._updateTakeoffMemo(_deltaTime);
         this._updateLandingMemo(_deltaTime);
-        this._updateAltitudeWarning();
+
+        if (this.updateThrottler.canUpdate(_deltaTime) !== -1) {
+            this._autopilotDisconnect(_deltaTime);
```
Comment (beheh) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:90]: I think the autopilot disconnect can probably be outside of this, 500ms is a long time before the warning appears.

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:90]: Ok

**Review by beheh — APPROVED** [review]

**Comment by beheh** [comment]
@Kimbyeoungjang Could you still add testing instructions to your opening comment please? Then the QA testers know what to do.

**Comment by Kimbyeoungjang** [comment]
> @Kimbyeoungjang Could you still add testing instructions to your opening comment please? Then the QA testers know what to do.
> 
> Looks good otherwise!

I added that :) thank you!

**Comment by darkfury334** [comment]
Category: QA Trial
Name: DarkFury#4936
Date of testing: Current date 18/03/21
Version of the sim: Version the tester uses
PR Tested: #3824
Tier of testing: 2
Changes to observe:
when Autopilot & Autothrottle disconnected, ECAM warning & aural warning will ringing.

Also add some master warning / master caution

Testing technique: try out and check

Steps you've made to test the Pull Request:
1. Spawned on RWY 33 at EDDH
2. Takeoff and climbed to 3000ft
3. Engaged A/T and A/P
4. Disengage A/T, check Upper ECAM and  Master warning button
5. Disengage A/P, check Upper ECAM and  Master warning button, listen for beep
6. Descend to 700ft without landing gear down and check ECAM warning & ringing.
7. Clear ECAM then retract flaps and check ECAM warning & ringing.
8. Clear ECAM and turn engine 1 off and check warnings
9. Start engine 1 and repeat step 8 for engine 2.
10. repeat steps 6 - 9 by following ecam instruction/ warning.

Flight notes (When done a full flight): N/A

Media: 
Microsoft Flight Simulator Screenshot 2021 03 18 - 11 50 35 12

Microsoft Flight Simulator Screenshot 2021 03 18 - 11 49 51 17

Microsoft Flight Simulator Screenshot 2021 03 18 - 12 57 34 92

Issues (If Any): No warning with flaps retracted below 750ft landing gear up.
Descend below 750ft and retract flaps. No warning on ECAM/Master and no ringing.  Was not on approach mode. See screenshot 3.

Overall Rating: Good
Conclusions of the testing: Overall good and works. had few issues as explained. Not sure if isolated case or if did something wrong

**Comment by Kimbyeoungjang** [comment]
thank you. let me check again and try to fix it.
i think descend below first, warning is on, and not climb or clear conflict lead to bug.
i'll fix in 5 min. thank you for your testing!

**Comment by Kimbyeoungjang** [comment]
> Category: QA Trial
> Name: DarkFury#4936
> Date of testing: Current date 18/03/21
> Version of the sim: Version the tester uses
> PR Tested: #3824
> Tier of testing: 2
> Changes to observe:
> when Autopilot & Autothrottle disconnected, ECAM warning & aural warning will ringing.
> 
> Also add some master warning / master caution
> 
> Testing technique: try out and check
> 
> Steps you've made to test the Pull Request:
> 
> 1. Spawned on RWY 33 at EDDH
> 2. Takeoff and climbed to 3000ft
> 3. Engaged A/T and A/P
> 4. Disengage A/T, check Upper ECAM and  Master warning button
> 5. Disengage A/P, check Upper ECAM and  Master warning button, listen for beep
> 6. Descend to 700ft without landing gear down and check ECAM warning & ringing.
> 7. Clear ECAM then retract flaps and check ECAM warning & ringing.
> 8. Clear ECAM and turn engine 1 off and check warnings
> 9. Start engine 1 and repeat step 8 for engine 2.
> 10. repeat steps 6 - 9 by following ecam instruction/ warning.
> 
> Flight notes (When done a full flight): N/A
> 
> Media:
> Microsoft Flight Simulator Screenshot 2021 03 18 - 11 50 35 12
> 
> Microsoft Flight Simulator Screenshot 2021 03 18 - 11 49 51 17
> 
> Microsoft Flight Simulator Screenshot 2021 03 18 - 12 57 34 92
> 
> Issues (If Any): No warning with flaps retracted below 750ft landing gear up not in toga.
> Descend below 750ft and retract flaps. No warning on ECAM/Master and no ringing. Was not on approach mode. See screenshot 3.
> 
> Overall Rating: Good
> Conclusions of the testing: Overall good and works. had few issues as explained. Not sure if isolated case or if did something wrong

test instruction changed and change some code.

can you test again with clear of conflict? if you have L/G NOT DOWN Warning, climb again more than 750ft(radio height) and change situation and try again please

**Comment by darkfury334** [comment]
Category: QA Trial
Name: DarkFury#4936
Date of testing: Current date 19/03/21
Version of the sim: Version the tester uses
PR Tested: #3824
Tier of testing: 2
Changes to observe:
when Autopilot & Autothrottle disconnected, ECAM warning & aural warning will ringing.

Also add some master warning / master caution

Testing technique: try out and check

Steps you've made to test the Pull Request:

1.Spawned on RWY 33 at EDDH
2.Takeoff and climbed to 2000ft
3.Engaged A/T and A/P
4.Disengage A/T, check Upper ECAM and Master warning button
5.Disengage A/P, check Upper ECAM and Master warning button, listen for beep
6.Descend to 600ft without landing gear down and check ECAM warning & ringing.
7.Climb back to 1000ft, deploy then retract flaps, steadily descend below 750ft (radio) and check ECAM warning & ringing.
8.Climb back to 1000ft, deploy Flaps, steadily descend below 750ft (radio), retract flaps and check ECAM warning & ringing.
9.Climb back to 1000ft, turn engine 1 off, steadily descend below 750ft (radio) and N1.
10.Start engine 1 and repeat step 8 for engine 2.


Flight notes (When done a full flight):N/A
Issues (If Any): N/A

Overall Rating: Good
Conclusions of the testing: After changing testing steps to clear conflict as requested, and changes made to code. it all seems to work well now

**Comment by St54Kevin** [comment]
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 21/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3824 
Tier of testing: 2
Changes to observe: ECAM warnings AP/ATHR

**Steps you've made to test the Pull Request:**
1. Loaded onto RW
2. Takeoff and climb on the SID
3. Disconnect AP via AP Inst. Disc. Button
4. Disconnect ATHR via ATHR Inst. Disc. Button
5. Descent to 500ft without LG down
6. Climb above 1000ft, descent back down with 1 ENG OFF and flaps

**Media:**

**Issues (If Any):** 
- Why is there a Triple Click after the AP disconnect? We dont have a mode reversion? Is this sound faked?
- Pressing the AP Inst. Disc. Button does not stop the Audio and lights. According to very first reference doc, it should.
- This case is not implemented at all. Should be achievable via FCU AP button ("by other means"):
image
- Pressing the ATHR Inst. Disc. Button does not stop the Audio and lights. According to very first reference doc, it should.
- Also this ATHR case is not implemented
image


**Conclusions of the testing:** Needs improvements. Cases in the reference doc are not covered at all. Remember the buttons on the FCU are NOT the instinctive disco buttons. Those are on the Joystick and Throttles. Also I dont like the triple click being faked, rather than connected to a real mode reversion

**Comment by Kimbyeoungjang** [comment]
> Category: QA Tester
> Name: St54Kevin#3583
> Date of testing: 21/03/2021
> Version of the sim: 1.14.5.0
> PR Tested: #3824
> Tier of testing: 2
> Changes to observe: ECAM warnings AP/ATHR
> 
> **Steps you've made to test the Pull Request:**
> 
> 1. Loaded onto RW
> 2. Takeoff and climb on the SID
> 3. Disconnect AP via AP Inst. Disc. Button
> 4. Disconnect ATHR via ATHR Inst. Disc. Button
> 5. Descent to 500ft without LG down
> 6. Climb above 1000ft, descent back down with 1 ENG OFF and flaps
> 
> **Media:**
> 
> **Issues (If Any):**
> 
> * Why is there a Triple Click after the AP disconnect? We dont have a mode reversion? Is this sound faked?
> * Pressing the AP Inst. Disc. Button does not stop the Audio and lights. According to very first reference doc, it should.
> * This case is not implemented at all. Should be achievable via FCU AP button ("by other means"):
>   image
> * Pressing the ATHR Inst. Disc. Button does not stop the Audio and lights. According to very first reference doc, it should.
> * Also this ATHR case is not implemented
>   image
> 
> **Conclusions of the testing:** Needs improvements. Cases in the reference doc are not covered at all. Remember the buttons on the FCU are NOT the instinctive disco buttons. Those are on the Joystick and Throttles. Also I dont like the triple click being faked, rather than connected to a real mode reversion

sorry. I'm not modify any sound source. that just default ap disconnect sound.

I'll remove that triple click!

also that document is old. i'll add newer :)

thank you for your feedback!

**Comment by St54Kevin** [comment]
> 
> sorry. I'm not modify any sound source. that just default ap disconnect sound.
> 
> I'll remove that triple click!
> 
> also that document is old. i'll add newer :)
> 
> thank you for your feedback!

This document was provided by ME and I can assure you its up to date on current Neo

**Pull request closed without merging**